add default None -> UserWarning

lilyminium opened the pull request:
Fixes #4746

Changes made in this Pull Request:
 - This mimics the Python warnings library by setting `category=None` to `UserWarning` by default

We don't appear to test our utilities anywhere so I haven't added a formal test -- but running it with #4744 worked.

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4747.org.readthedocs.build/en/4747/

<!-- readthedocs-preview mdanalysis end -->

lilyminium committed:
add default None -> UserWarning

lilyminium committed:
add comment

tylerjereddy commented on testsuite/MDAnalysisTests/util.py:
```diff
@@ -233,6 +233,11 @@ def _warn(self, message, category=None, *args, **kwargs):
+            # https://docs.python.org/3/library/warnings.html#warnings.warn
+            # if category is None, the default UserWarning is used
+            if category is None:
+                category = UserWarning
```
Ah, this is a surprisingly simple fix, very nice. Confirming it also works in my hands.

I'll approve but not merge yet given the release cycle situation.

IAlibay reviewed (changes requested):
It would be best if this was applied to something in our tests - i.e. some kind of testing by application.

orbeckst commented:
@tylerjereddy can you keep an eye on this PR, please? If you don't have the bandwidth, please unassign yourself again. Thanks.

tylerjereddy commented:
@orbeckst @lilyminium testing locally, we could just follow through with our original intention here and the `develop` branch will fail this test without the patch that Lily is providing here. This fails on `develop` and passes in this branch:

```diff
--- a/testsuite/MDAnalysisTests/topology/test_itp.py
+++ b/testsuite/MDAnalysisTests/topology/test_itp.py
@@ -26,6 +26,7 @@ import MDAnalysis as mda
 import numpy as np
 from numpy.testing import assert_allclose, assert_equal
 
+from MDAnalysisTests.util import no_deprecated_call
 from MDAnalysisTests.topology.base import ParserBase
 from MDAnalysisTests.datafiles import (
     ITP,  # GROMACS itp
@@ -490,3 +491,5 @@ def test_missing_elements_no_attribute():
         u = mda.Universe(ITP_atomtypes)
     with pytest.raises(AttributeError):
         _ = u.atoms.elements
+    with no_deprecated_call():
+        mda.Universe(ITP_atomtypes)
```

That's what we meant to do all along anyway, and should match Irfan's request as well--actually using the functionality to test it.

orbeckst commented:
Sounds good @tylerjereddy !

tylerjereddy committed:
TST: PR 4747 revisions

* Add a regression test that serves the dual purpose of ensuring
that the `no_deprecated_call` decorator behaves properly when
the warning category is `None`, and also enforces our intended
warnings behavior for the element guessing changes in MDAnalysisgh-4744.

tylerjereddy commented:
I rebased and pushed in my suggested adjustment.

IAlibay reviewed (approved):
Thanks, lgtm!

IAlibay commented:
Looks like failures are related to the rdkit bug, so this should still be good to go.

RMeli commented:
@IAlibay are you suggesting to merge despite the failures?

IAlibay commented:
> @IAlibay are you suggesting to merge despite the failures?

If folks don't mind then yes, we'll need to fix the rdkit failure at some point but that's probably not happening today.

tylerjereddy commented:
Ok, I squash-merged given the exact same CI failures are seen in other recent unrelated PRs.

Technically I pushed in a commit here so that's a bit like self-merging, but obviously the changes are very small, not really user-facing, and the approval was from another core dev.
